File: app/transactions.py
"""Transaction and review inbox read models."""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import date
from decimal import Decimal, InvalidOperation
from typing import LiteralString, cast

import psycopg


@dataclass(frozen=True)
class TransactionListItem:
    transaction_id: int
    booking_date: date
    amount: str
    merchant: str
    description: str
    category: str
    classification_method: str
    needs_review: bool
    flags: tuple[str, ...]


@dataclass(frozen=True)
class RawTransactionDetails:
    transaction_id: int
    booking_date: date
    amount: str
    merchant: str
    description: str
    category: str
    details: tuple[tuple[str, str], ...]
    payload_json: str


@dataclass(frozen=True)
class TransactionFilters:
    query: str = ""
    month: str = ""
    category: str = ""
    min_amount: str = ""
    max_amount: str = ""
    kind: str = ""
    needs_review: bool | None = None

    @property
    def has_any(self) -> bool:
        return bool(
            self.query
            or self.month
            or self.category
            or self.min_amount
            or self.max_amount
            or self.kind
            or self.needs_review is not None
        )

    @property
    def has_advanced(self) -> bool:
        return bool(self.month or self.category or self.min_amount or self.max_amount or self.kind or self.needs_review)


def list_transactions(
    database_url: str,
    *,
    needs_review: bool | None = None,
    filters: TransactionFilters | None = None,
) -> list[TransactionListItem]:
    active_filters = filters or TransactionFilters(needs_review=needs_review)
    if needs_review is not None:
        active_filters = TransactionFilters(
            query=active_filters.query,
            month=active_filters.month,
            category=active_filters.category,
            min_amount=active_filters.min_amount,
            max_amount=active_filters.max_amount,
            kind=active_filters.kind,
            needs_review=needs_review,
        )
    where_clause, params = _build_where_clause(active_filters)
    with psycopg.connect(database_url) as connection:
        rows = connection.execute(
            cast(
                LiteralString,
                f"""
            SELECT
                enriched_transactions.id,
                raw_transactions.booking_date,
                raw_transactions.amount,
                COALESCE(merchants.name, raw_transactions.counterparty_name, 'Unknown'),
                raw_transactions.description,
                COALESCE(categories.name, 'Unknown'),
                COALESCE(enriched_transactions.classification_method, 'none'),
                enriched_transactions.needs_review,
                enriched_transactions.is_income,
                enriched_transactions.is_transfer,
                enriched_transactions.is_savings,
                enriched_transactions.is_fixed_cost,
                enriched_transactions.is_recurring,
                enriched_transactions.is_one_off,
                enriched_transactions.is_excluded_from_budget
            FROM enriched_transactions
            JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
            LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
            LEFT JOIN categories ON categories.id = enriched_transactions.category_id
            {where_clause}
            ORDER BY raw_transactions.booking_date DESC, enriched_transactions.id DESC
            LIMIT 100
            """,
            ),
            tuple(params),
        ).fetchall()

    return [_row_to_transaction(row) for row in rows]


def get_transaction_raw_details(database_url: str, transaction_id: int) -> RawTransactionDetails:
    with psycopg.connect(database_url) as connection:
        row = connection.execute(
            """
            SELECT
                enriched_transactions.id,
                raw_transactions.booking_date,
                raw_transactions.amount,
                COALESCE(merchants.name, raw_transactions.counterparty_name, 'Unknown'),
                raw_transactions.description,
                COALESCE(categories.name, 'Unknown'),
                accounts.name,
                accounts.iban,
                raw_transactions.provider,
                raw_transactions.provider_transaction_id,
                raw_transactions.source_hash,
                raw_transactions.value_date,
                raw_transactions.currency,
                raw_transactions.counterparty_name,
                raw_transactions.counterparty_iban,
                raw_transactions.raw_payload_json,
                raw_transactions.first_seen_at,
                raw_transactions.last_seen_at
            FROM enriched_transactions
            JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
            JOIN accounts ON accounts.id = raw_transactions.account_id
            LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
            LEFT JOIN categories ON categories.id = enriched_transactions.category_id
            WHERE enriched_transactions.id = %s
            """,
            (transaction_id,),
        ).fetchone()
    if row is None:
        raise ValueError(f"transaction not found: {transaction_id}")
    return _row_to_raw_details(row)


def _build_where_clause(filters: TransactionFilters) -> tuple[str, list[object]]:
    clauses: list[str] = []
    params: list[object] = []
    if filters.needs_review is not None:
        clauses.append("enriched_transactions.needs_review = %s")
        params.append(filters.needs_review)
    if filters.query:
        clauses.append(
            """
            (
                raw_transactions.description ILIKE %s
                OR raw_transactions.counterparty_name ILIKE %s
                OR merchants.name ILIKE %s
            )
            """
        )
        like_value = f"%{filters.query}%"
        params.extend([like_value, like_value, like_value])
    if filters.month:
        clauses.append("to_char(raw_transactions.booking_date, 'YYYY-MM') = %s")
        params.append(filters.month)
    if filters.category:
        clauses.append("categories.name = %s")
        params.append(filters.category)
    min_amount = _parse_amount_filter(filters.min_amount)
    if min_amount is not None:
        clauses.append("ABS(raw_transactions.amount) >= %s")
        params.append(min_amount)
    elif filters.min_amount:
        clauses.append("FALSE")
    max_amount = _parse_amount_filter(filters.max_amount)
    if max_amount is not None:
        clauses.append("ABS(raw_transactions.amount) <= %s")
        params.append(max_amount)
    elif filters.max_amount:
        clauses.append("FALSE")
    kind_clause = _kind_clause(filters.kind)
    if kind_clause:
        clauses.append(kind_clause)
    if not clauses:
        return "", params
    return "WHERE " + " AND ".join(clauses), params


def _parse_amount_filter(value: str) -> Decimal | None:
    if not value:
        return None
    try:
        return abs(Decimal(value.strip().replace(",", "")))
    except InvalidOperation:
        return None


def _kind_clause(kind: str) -> str:
    return {
        "income": "enriched_transactions.is_income",
        "spending": (
            "raw_transactions.amount < 0 "
            "AND NOT enriched_transactions.is_transfer "
            "AND NOT enriched_transactions.is_excluded_from_budget"
        ),
        "transfer": "enriched_transactions.is_transfer",
        "savings": "enriched_transactions.is_savings",
        "fixed": "enriched_transactions.is_fixed_cost",
        "recurring": "enriched_transactions.is_recurring",
        "one_off": "enriched_transactions.is_one_off",
        "excluded": "enriched_transactions.is_excluded_from_budget",
    }.get(kind, "")


def _row_to_transaction(row: tuple[object, ...]) -> TransactionListItem:
    return TransactionListItem(
        transaction_id=_expect_int(row[0]),
        booking_date=_expect_date(row[1]),
        amount=_format_money(row[2]),
        merchant=str(row[3]),
        description=str(row[4]),
        category=str(row[5]),
        classification_method=str(row[6]),
        needs_review=bool(row[7]),
        flags=_flags(row),
    )


def _row_to_raw_details(row: tuple[object, ...]) -> RawTransactionDetails:
    payload = row[15] if row[15] is not None else {}
    return RawTransactionDetails(
        transaction_id=_expect_int(row[0]),
        booking_date=_expect_date(row[1]),
        amount=_format_money(row[2]),
        merchant=str(row[3]),
        description=str(row[4]),
        category=str(row[5]),
        details=(
            ("Account", str(row[6])),
            ("IBAN", _display_optional(row[7])),
            ("Provider", str(row[8])),
            ("Provider transaction ID", _display_optional(row[9])),
            ("Source hash", str(row[10])),
            ("Value date", _display_optional(row[11])),
            ("Currency", str(row[12])),
            ("Counterparty", _display_optional(row[13])),
            ("Counterparty IBAN", _display_optional(row[14])),
            ("First seen", str(row[16])),
            ("Last seen", str(row[17])),
        ),
        payload_json=json.dumps(payload, indent=2, sort_keys=True),
    )


def _flags(row: tuple[object, ...]) -> tuple[str, ...]:
    flags: list[str] = []
    flag_columns = (
        ("income", row[8]),
        ("transfer", row[9]),
        ("savings", row[10]),
        ("fixed", row[11]),
        ("recurring", row[12]),
        ("one-off", row[13]),
        ("excluded", row[14]),
    )
    for label, enabled in flag_columns:
        if enabled:
            flags.append(label)
    return tuple(flags)


def _format_money(value: object) -> str:
    amount = Decimal(str(value)).quantize(Decimal("0.01"))
    sign = "-" if amount < 0 else ""
    absolute = abs(amount)
    return f"{sign}EUR {absolute:,.2f}"


def _expect_int(value: object) -> int:
    if not isinstance(value, int):
        raise RuntimeError(f"expected integer id, got {type(value).__name__}")
    return value


def _expect_date(value: object) -> date:
    if not isinstance(value, date):
        raise RuntimeError(f"expected date, got {type(value).__name__}")
    return value


def _display_optional(value: object) -> str:
    if value is None or value == "":
        return "None"
    return str(value)

File: src/server/management/transactions.ts
import type pg from "pg";

import type {
  TransactionFilters,
  TransactionRawDetails,
  TransactionsResponse,
  TransactionUpdateRequest,
} from "../../shared/management.ts";
import { sql, toQuery } from "../db/sql.ts";
import { listCategories } from "./service.ts";

export type TransactionRow = {
  amount: string;
  booking_date: string;
  category_id: string | null;
  category_name: string | null;
  classification_method: string | null;
  description: string;
  id: string;
  is_excluded_from_budget: boolean;
  is_fixed_cost: boolean;
  is_income: boolean;
  is_one_off: boolean;
  is_recurring: boolean;
  is_savings: boolean;
  is_transfer: boolean;
  merchant: string | null;
  needs_review: boolean;
};

export type TransactionMatchRow = TransactionRow & {
  category_matches: boolean;
  has_manual_override: boolean;
  merchant_matches: boolean;
};

export async function listTransactions(
  pool: pg.Pool,
  filters: TransactionFilters = {},
): Promise<TransactionsResponse> {
  const where = buildTransactionWhere(filters);
  const [categories, merchants, transactions] = await Promise.all([
    listCategories(pool),
    listMerchantNames(pool),
    pool.query<TransactionRow>(
      `
        SELECT
          enriched_transactions.id,
          raw_transactions.booking_date::text,
          raw_transactions.amount::text,
          COALESCE(merchants.name, raw_transactions.counterparty_name, 'Unknown') AS merchant,
          raw_transactions.description,
          categories.id::text AS category_id,
          categories.name AS category_name,
          COALESCE(enriched_transactions.classification_method, 'none') AS classification_method,
          enriched_transactions.needs_review,
          enriched_transactions.is_income,
          enriched_transactions.is_transfer,
          enriched_transactions.is_savings,
          enriched_transactions.is_fixed_cost,
          enriched_transactions.is_recurring,
          enriched_transactions.is_one_off,
          enriched_transactions.is_excluded_from_budget
        FROM enriched_transactions
        JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
        LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
        LEFT JOIN categories ON categories.id = enriched_transactions.category_id
        ${where.clause}
        ORDER BY raw_transactions.booking_date DESC, enriched_transactions.id DESC
        LIMIT 100
      `,
      where.values,
    ),
  ]);

  return {
    categories,
    merchants,
    transactions: transactions.rows.map(transactionRow),
  };
}

export async function getTransactionRawDetails(
  pool: pg.Pool,
  transactionId: number,
): Promise<TransactionRawDetails> {
  const result = await pool.query<{
    account_iban: string | null;
    account_name: string;
    amount: string;
    booking_date: string;
    category_name: string | null;
    counterparty_iban: string | null;
    counterparty_name: string | null;
    currency: string;
    description: string;
    first_seen_at: string;
    id: string;
    last_seen_at: string;
    merchant: string | null;
    provider: string;
    provider_transaction_id: string | null;
    raw_payload_json: unknown;
    source_hash: string;
    value_date: string | null;
  }>(
    `
      SELECT
        enriched_transactions.id,
        raw_transactions.booking_date::text,
        raw_transactions.amount::text,
        COALESCE(merchants.name, raw_transactions.counterparty_name, 'Unknown') AS merchant,
        raw_transactions.description,
        categories.name AS category_name,
        accounts.name AS account_name,
        accounts.iban AS account_iban,
        raw_transactions.provider,
        raw_transactions.provider_transaction_id,
        raw_transactions.source_hash,
        raw_transactions.value_date::text,
        raw_transactions.currency,
        raw_transactions.counterparty_name,
        raw_transactions.counterparty_iban,
        raw_transactions.raw_payload_json,
        raw_transactions.first_seen_at::text,
        raw_transactions.last_seen_at::text
      FROM enriched_transactions
      JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
      JOIN accounts ON accounts.id = raw_transactions.account_id
      LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
      LEFT JOIN categories ON categories.id = enriched_transactions.category_id
      WHERE enriched_transactions.id = $1
    `,
    [transactionId],
  );
  const row = result.rows[0];
  if (!row) {
    throw new Error("Transaction not found");
  }

  return {
    amount: row.amount,
    bookingDate: row.booking_date,
    categoryName: row.category_name,
    description: row.description,
    details: [
      { label: "Account", value: row.account_name },
      { label: "IBAN", value: displayValue(row.account_iban) },
      { label: "Provider", value: row.provider },
      { label: "Provider transaction ID", value: displayValue(row.provider_transaction_id) },
      { label: "Source hash", value: row.source_hash },
      { label: "Value date", value: displayValue(row.value_date) },
      { label: "Currency", value: row.currency },
      { label: "Counterparty", value: displayValue(row.counterparty_name) },
      { label: "Counterparty IBAN", value: displayValue(row.counterparty_iban) },
      { label: "First seen", value: row.first_seen_at },
      { label: "Last seen", value: row.last_seen_at },
    ],
    id: Number(row.id),
    merchant: row.merchant ?? "Unknown",
    payloadJson: JSON.stringify(row.raw_payload_json ?? {}, null, 2),
  };
}

export async function updateTransaction(
  pool: pg.Pool,
  transactionId: number,
  update: TransactionUpdateRequest,
): Promise<void> {
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    const merchantId = await upsertMerchant(client, update.merchantName, update.categoryId);
    await upsertManualOverride(client, transactionId, update, merchantId);
    if (update.createAlias && merchantId !== null) {
      await upsertMerchantAlias(client, transactionId, merchantId);
    }
    await client.query(
      toQuery(sql`
        UPDATE enriched_transactions
        SET
          category_id = ${update.categoryId},
          merchant_id = ${merchantId},
          is_transfer = ${update.flags.isTransfer},
          is_savings = ${update.flags.isSavings},
          is_one_off = ${update.flags.isOneOff},
          is_excluded_from_budget = ${update.flags.isExcludedFromBudget},
          is_variable_cost = NOT (${update.flags.isTransfer} OR ${update.flags.isSavings}),
          needs_review = false,
          classification_method = 'manual_override',
          classification_confidence = 1,
          classification_reason = 'Edited from transaction workspace.',
          classification_model = NULL,
          classification_runtime = NULL,
          classification_prompt_version = NULL,
          updated_at = now()
        WHERE id = ${transactionId}
      `),
    );
    await client.query("COMMIT");
  } catch (error) {
    await client.query("ROLLBACK");
    throw error;
  } finally {
    client.release();
  }
}

async function listMerchantNames(pool: pg.Pool): Promise<string[]> {
  const result = await pool.query<{ merchant: string }>(`
    SELECT DISTINCT COALESCE(merchants.name, raw_transactions.counterparty_name) AS merchant
    FROM enriched_transactions
    JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
    LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
    WHERE COALESCE(merchants.name, raw_transactions.counterparty_name) IS NOT NULL
      AND COALESCE(merchants.name, raw_transactions.counterparty_name) <> ''
    ORDER BY merchant
  `);
  return result.rows.map((row) => row.merchant);
}

async function upsertMerchant(
  client: pg.PoolClient,
  merchantName: string | undefined,
  categoryId: number,
): Promise<number | null> {
  const cleaned = merchantName?.trim();
  if (!cleaned) {
    return null;
  }
  const result = await client.query<{ id: string }>(
    toQuery(sql`
      INSERT INTO merchants (name, normalized_name, default_category_id)
      VALUES (${cleaned}, ${cleaned.toLowerCase()}, ${categoryId})
      ON CONFLICT (normalized_name)
      DO UPDATE SET default_category_id = EXCLUDED.default_category_id, updated_at = now()
      RETURNING id
    `),
  );
  return Number(result.rows[0]?.id);
}

async function upsertManualOverride(
  client: pg.PoolClient,
  transactionId: number,
  update: TransactionUpdateRequest,
  merchantId: number | null,
): Promise<void> {
  await client.query(
    toQuery(sql`
      INSERT INTO manual_overrides (enriched_transaction_id, category_id, merchant_id, flags_json)
      VALUES (
        ${transactionId},
        ${update.categoryId},
        ${merchantId},
        ${JSON.stringify({
          is_excluded_from_budget: update.flags.isExcludedFromBudget,
          is_one_off: update.flags.isOneOff,
          is_savings: update.flags.isSavings,
          is_transfer: update.flags.isTransfer,
        })}::jsonb
      )
      ON CONFLICT (enriched_transaction_id)
      DO UPDATE SET
        category_id = EXCLUDED.category_id,
        merchant_id = EXCLUDED.merchant_id,
        flags_json = EXCLUDED.flags_json,
        updated_at = now()
    `),
  );
}

async function upsertMerchantAlias(
  client: pg.PoolClient,
  transactionId: number,
  merchantId: number,
): Promise<void> {
  const result = await client.query<{ match_text: string | null }>(
    `
      SELECT COALESCE(NULLIF(trim(raw_transactions.counterparty_name), ''), NULLIF(trim(raw_transactions.description), '')) AS match_text
      FROM enriched_transactions
      JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
      WHERE enriched_transactions.id = $1
    `,
    [transactionId],
  );
  const matchText = result.rows[0]?.match_text;
  if (!matchText) {
    return;
  }
  await client.query(
    toQuery(sql`
      INSERT INTO merchant_aliases (merchant_id, match_text, match_type, priority)
      SELECT ${merchantId}, ${matchText}, 'contains', 50
      WHERE NOT EXISTS (
        SELECT 1
        FROM merchant_aliases
        WHERE merchant_id = ${merchantId}
          AND match_text = ${matchText}
          AND match_type = 'contains'
      )
    `),
  );
}

function buildTransactionWhere(filters: TransactionFilters): { clause: string; values: unknown[] } {
  const clauses: string[] = [];
  const values: unknown[] = [];
  const add = (clause: string, value?: unknown) => {
    clauses.push(clause.replace("?", `$${values.length + 1}`));
    if (value !== undefined) {
      values.push(value);
    }
  };

  if (filters.needsReview !== undefined) {
    add("enriched_transactions.needs_review = ?", filters.needsReview);
  }
  if (filters.query) {
    const param = `$${values.length + 1}`;
    clauses.push(`(
      raw_transactions.description ILIKE '%' || ${param} || '%'
      OR raw_transactions.counterparty_name ILIKE '%' || ${param} || '%'
      OR merchants.name ILIKE '%' || ${param} || '%'
    )`);
    values.push(filters.query);
  }
  if (filters.month) {
    add("to_char(raw_transactions.booking_date, 'YYYY-MM') = ?", filters.month);
  }
  if (filters.category) {
    add("categories.name = ?", filters.category);
  }
  if (filters.merchant) {
    add("COALESCE(merchants.name, raw_transactions.counterparty_name) = ?", filters.merchant);
  }
  addAmountClause("ABS(raw_transactions.amount) >= ?", filters.minAmount, clauses, values);
  addAmountClause("ABS(raw_transactions.amount) <= ?", filters.maxAmount, clauses, values);
  const kindClause = kindFilter(filters.kind);
  if (kindClause) {
    clauses.push(kindClause);
  }

  return { clause: clauses.length ? `WHERE ${clauses.join(" AND ")}` : "", values };
}

function addAmountClause(
  clause: string,
  amount: string | undefined,
  clauses: string[],
  values: unknown[],
): void {
  if (!amount) {
    return;
  }
  const parsed = Number(amount.replace(",", ""));
  if (!Number.isFinite(parsed)) {
    clauses.push("FALSE");
    return;
  }
  clauses.push(clause.replace("?", `$${values.length + 1}`));
  values.push(Math.abs(parsed));
}

function kindFilter(kind: string | undefined): string {
  switch (kind) {
    case "excluded":
      return "enriched_transactions.is_excluded_from_budget";
    case "fixed":
      return "enriched_transactions.is_fixed_cost";
    case "income":
      return "enriched_transactions.is_income";
    case "one_off":
      return "enriched_transactions.is_one_off";
    case "recurring":
      return "enriched_transactions.is_recurring";
    case "savings":
      return "enriched_transactions.is_savings";
    case "spending":
      return "raw_transactions.amount < 0 AND NOT enriched_transactions.is_transfer AND NOT enriched_transactions.is_excluded_from_budget";
    case "transfer":
      return "enriched_transactions.is_transfer";
    case "uncategorized":
      return "(categories.name IS NULL OR categories.name = 'Unknown')";
    default:
      return "";
  }
}

export function transactionRow(row: TransactionRow) {
  return {
    amount: row.amount,
    bookingDate: row.booking_date,
    categoryId: row.category_id ? Number(row.category_id) : null,
    categoryName: row.category_name,
    classificationMethod: row.classification_method,
    description: row.description,
    flags: flags(row),
    id: Number(row.id),
    merchant: row.merchant ?? "Unknown",
    needsReview: row.needs_review,
  };
}

function flags(row: TransactionRow): string[] {
  return [
    row.is_income ? "income" : null,
    row.is_transfer ? "transfer" : null,
    row.is_savings ? "savings" : null,
    row.is_fixed_cost ? "fixed" : null,
    row.is_recurring ? "recurring" : null,
    row.is_one_off ? "one-off" : null,
    row.is_excluded_from_budget ? "excluded" : null,
  ].filter((flag): flag is string => flag !== null);
}

function displayValue(value: string | null): string {
  return value && value !== "" ? value : "None";
}
